src/compiler/checker.ts

```diff
@@ -15974,17 +15974,26 @@ namespace ts {
                         return grammarErrorOnFirstToken(variableList.declarations[1], diagnostic);
                     }
                     const firstDeclaration = variableList.declarations[0];
-                    if (firstDeclaration.initializer) {
-                        const diagnostic = forInOrOfStatement.kind === SyntaxKind.ForInStatement
-                            ? Diagnostics.The_variable_declaration_of_a_for_in_statement_cannot_have_an_initializer
-                            : Diagnostics.The_variable_declaration_of_a_for_of_statement_cannot_have_an_initializer;
-                        return grammarErrorOnNode(firstDeclaration.name, diagnostic);
-                    }
-                    if (firstDeclaration.type) {
-                        const diagnostic = forInOrOfStatement.kind === SyntaxKind.ForInStatement
-                            ? Diagnostics.The_left_hand_side_of_a_for_in_statement_cannot_use_a_type_annotation
-                            : Diagnostics.The_left_hand_side_of_a_for_of_statement_cannot_use_a_type_annotation;
-                        return grammarErrorOnNode(firstDeclaration, diagnostic);
+
+                    // firstDeclaration can be undefined if there is variable declaration in for-of or for-in
+                    // See http://www.ecma-international.org/ecma-262/6.0/#sec-for-in-and-for-of-statements for details
+                    // For example:
+                    //      var let = 10;
+                    //      for (let of [1,2,3]) {} // this is invalid ES6 syntax
+                    //      for (let in [1,2,3]) {} // this is invalid ES6 syntax
+                    if (firstDeclaration) {

```

Comment:
Consider making this a check on `variableList.declaration.length > 0`. @vladima might be able to weigh in on if this sort of thing causes de-opts.